src/hotspot/os/posix/signals_posix.cpp

```diff
@@ -502,21 +528,106 @@ void PosixSignals::unblock_error_signals() {
   ::pthread_sigmask(SIG_UNBLOCK, &set, NULL);
 }
 
-// Renamed from 'signalHandler' to avoid collision with other shared libs.
-static void javaSignalHandler(int sig, siginfo_t* info, void* uc) {
-  assert(info != NULL && uc != NULL, "it must be old kernel");
+class ErrnoPreserver: public StackObj {
+  const int _saved;
+public:
+  ErrnoPreserver() : _saved(errno) {}
+  ~ErrnoPreserver() { errno = _saved; }
+};
+
+// Entry point for the hotspot signal handler.
+static void javaSignalHandler(int sig, siginfo_t* info, void* ucVoid) {
+  // Do not add any code here. Add code to javaSignalHandler_inner.
+  (void)javaSignalHandler_inner(sig, info, ucVoid, true);
+}
+
+static bool javaSignalHandler_inner(int sig, siginfo_t* info, void* ucVoid, bool abort_if_unrecognized) {
+  assert(info != NULL && ucVoid != NULL, "sanity");
 
+  // Note: it's not uncommon that JNI code uses signal/sigset to install,
+  // then restore certain signal handler (e.g. to temporarily block SIGPIPE,
+  // or have a SIGILL handler when detecting CPU type). When that happens,
+  // this handler might be invoked with junk info/ucVoid. To avoid unnecessary
+  // crash when libjsig is not preloaded, try handle signals that do not require
+  // siginfo/ucontext first.
+
+  // Preserve errno value over signal handler.
+  //  (note: RAII ok here, even with JFR thread crash protection, see below).
+  ErrnoPreserver ep;
+
+  // Unblock all synchronous error signals (see JDK-8252533)
   PosixSignals::unblock_error_signals();
 
-  int orig_errno = errno;  // Preserve errno value over signal handler.
-#if defined(BSD)
-  JVM_handle_bsd_signal(sig, info, uc, true);
-#elif defined(AIX)
-  JVM_handle_aix_signal(sig, info, uc, true);
+  ucontext_t* const uc = (ucontext_t*) ucVoid;
+  Thread* const t = Thread::current_or_null_safe();
+  JavaThread* const jt = (t != NULL && t->is_Java_thread()) ? (JavaThread*) t : NULL;
+
+  // Handle JFR thread crash protection.
+  //  Note: this may cause us to longjmp away. Do not use any code before this
+  //  point which really needs any form of epilogue code running, eg RAII objects.
+  os::ThreadCrashProtection::check_crash_protection(sig, t);
+
+  bool signal_was_handled = false;
+
+  // Handle assertion poison page accesses.
+#ifdef CAN_SHOW_REGISTERS_ON_ASSERT
+  if ((sig == SIGSEGV || sig == SIGBUS) && info != NULL && info->si_addr == g_assert_poison) {
+    signal_was_handled = handle_assert_poison_fault(ucVoid, info->si_addr);
+  }
+#endif
+
+  // Ignore SIGPIPE and SIGXFSZ (4229104, 6499219).
+  if (sig == SIGPIPE || sig == SIGXFSZ) {
+    PosixSignals::chained_handler(sig, info, ucVoid);
+    signal_was_handled = true; // unconditionally.
+  }
+
+  // Call platform dependent signal handler.
+  if (!signal_was_handled) {
+    signal_was_handled = PosixSignals::pd_hotspot_signal_handler(sig, info, uc, jt);
```

Comment:
Can we move:

JavaThread* const jt = (t != NULL && t->is_Java_thread()) ? (JavaThread*) t : NULL;

here? It's not used anywhere else.

Reply:
Sure, I can do this. We can move it up again when/if later needed.